web/index.html

```diff
@@ -552,64 +601,147 @@ <h3>Used in</h3>
+                                    if (!usedElements.has(this.data.data[currentElement].from[i][1])) {
+                                        fromIndex = i;
+                                        break;
+                                    }
+                                } else if (this.saveData.has(this.data.data[currentElement].from[i][1])) {
+                                    if (!usedElements.has(this.data.data[currentElement].from[i][0])) {
+                                        fromIndex = i;
+                                        break;
+                                    }
                                 } else {
-                                    hiddenTo = hiddenTo.split(sep2).map(x => x.split(sep3))
+                                    fromIndex = i;
+                                    break;
                                 }
-                                return [x[0], { from, to, hiddenFrom, hiddenTo }]
-                            }))
-                            this.data.data = Object.freeze(Object.assign({}, this.data.data, chunkData))
-                            this.data.metadata.loadedChunks.push(chunk)
-                        });
+                            }
+
+                            if (fromIndex == null) {
+                                console.log('Added  looping element: ' + this.data.index[currentElement][1]);
+                                // If a dependency loop is detected move to the parent again and try the following recipe instead
+                                usedElements.delete(currentNode.parent.value);
+                                craftingList.unshift(currentNode.parent);
+                                continue;
+                            } else {
+                                currentNode.left = { value: this.data.data[currentElement].from[fromIndex][0], parent: currentNode };
+                                currentNode.right = { value: this.data.data[currentElement].from[fromIndex][1], parent: currentNode };
+                                currentNode.lastIndex = fromIndex;
+                                usedElements.add(currentElement);
+                            }
+                        }
+
+                        result.unshift({ A: currentNode.left.value, B: currentNode.right.value, C: currentElement });
+                        if (this.data.costs[currentNode.left.value] != 1 && (!this.saveData.has(currentNode.left.value) || this.showFullRecipe)) {
+                            craftingList.unshift(currentNode.left);
+                        }
+                        if (this.data.costs[currentNode.right.value] != 1 && (!this.saveData.has(currentNode.right.value) || this.showFullRecipe)) {
+                            craftingList.unshift(currentNode.right);
+                        }
+                    }
+                    this.craftingTreeData = result;
+                },
+                loadChunk: async function (chunk) {
```

Comment:
Load chunk is pretty straightforward, we first check if the current chunk is being loaded or has already been loaded, if so return the previous result. The return value is of type Promise<bool>, which can be awaited on nicely.
Next we setup a worker using the CHUNK_WORKER_PATH, which is just loadChunkWorker.js.
Then we setup some handlers for when it is done or if it hits an error. And then we start it using a post message by sending it the data it should use to run.